### ISSUE TITLE
How to make heatmap for model with VGG16 base and new classifier

### ISSUE THREAD

**Issue by MirandaTilton** [rstudio/keras#581]
I have trained a new classifier on the VGG16 base and I am now trying to create a heatmap to visualize which parts of the image are activated when classifying that image. (To create the heatmap, I am modifying code from Chapter 3 of _Deep Learning with R_ by Francois Challet with J.J. Allaire.) 

Attached is simple code for a reproducible example. The first half of the code creates a heatmap using the original classifier for VGG16. The second half trains a simple model and details where the heatmap process is no longer working.

https://gist.github.com/MirandaTilton/ff3404175ff0c05bb5113d4854dc39c3

`devtools::source_gist("https://gist.github.com/MirandaTilton/ff3404175ff0c05bb5113d4854dc39c3", filename = "cifar10example_heatmapgeneration.R")`

**Comment by srvanderplas** [comment]
Using @skeydan's code with a few modifications, I have your finished heatmap code looking something like this:

```
#### cifar10 example modified from
#### https://keras.rstudio.com/articles/examples/cifar10_cnn.html

#### Heatmap code modified from Ch 3 of Deep Learning with R
#### by Francois Challet with J.J. Allaire.
#### For more info, see https://www.manning.com/books/deep-learning-with-r

library(keras)

# Data Preparation --------------------------------------------------------

cifar10 <- dataset_cifar10()

# Feature scale RGB values in test and train inputs  
x_train <- cifar10$train$x/255
x_test <- cifar10$test$x/255
y_train <- to_categorical(cifar10$train$y, num_classes = 10)
y_test <- to_categorical(cifar10$test$y, num_classes = 10)

# Select target image and format for prediction
#### Pretrained classifier requires 224x224
#### New classifier will take image size as is (32x32)
img_32 <- x_test[1,,,]
dim(img_32) <- c(1, 32, 32, 3)
img_224 <- array(0, dim = c(1, 224, 224, 3))
img_224[1, 97:128, 97:128, ] <- img_32


# Heatmap using pre-trained classifier -------------------------------------

# Load VGG16 (including classifier)
VGG16 <- application_vgg16(
  include_top = TRUE,
  weights = "imagenet"
)

# Create and train model w/ VGG16 base, new classifier --------------------
#### Note: the training parameters in this example are optimized for speed
#### and not for classification accuracy

input <- layer_input(shape = c(32, 32, 3))

# Load VGG16 base (without classifier)
VGG16_base <- application_vgg16(
  include_top = FALSE,
  weights = "imagenet",
  input_tensor = input
)


output <- VGG16_base$output %>%
  layer_flatten(input_shape = layer_input(shape = c(32, 32, 512))) %>%
  layer_dense(units = 10, activation = "softmax")

model <- keras_model(input, output)

predictions <- model %>% predict(img_32)

img_output <- model$output[,1]

last_conv_layer <- model %>% get_layer("block5_conv3")
grads <- k_gradients(img_output, last_conv_layer$output)[[1]]


# finish making the heatmap
pooled_grads <- k_mean(grads, axis = c(1, 2, 3))
iterate <- k_function(list(VGG16$input),
                      list(pooled_grads, last_conv_layer$output[1,,,]))
c(pooled_grads_value, conv_layer_output_value) %<-% iterate(list(img_32))
for (i in 1:512) {
  conv_layer_output_value[,,i] <-
    conv_layer_output_value[,,i] * pooled_grads_value[[i]]
}
heatmap <- apply(conv_layer_output_value, c(1,2), mean)
heatmap <- pmax(heatmap, 0)
heatmap <- heatmap / max(heatmap)

plot_heatmap(heatmap, width = 32, height = 32)
```

which produced the following image: 
![image](https://user-images.githubusercontent.com/2127125/47661912-f39bf980-db67-11e8-8e51-e35339dbefce.png)

**Comment by srvanderplas** [comment]
We spent some time trying to figure out what this was doing and how to actually fit a model using frozen weights for the VGG16 portion. Here's our code:

```
library(keras)
plot_heatmap <- function(heatmap, width = 224, height = 224,
                         bg = "white", col = terrain.colors(12)) {
  op = par(mar = c(0,0,0,0))
  on.exit({par(op)}, add = TRUE)
  rotate <- function(x) t(apply(x, 2, rev))
  image(rotate(heatmap), axes = FALSE, asp = 1, col = col)
}


cifar10 <- dataset_cifar10()

# Feature scale RGB values in test and train inputs
x_train <- cifar10$train$x/255
x_test <- cifar10$test$x/255
y_train <- to_categorical(cifar10$train$y, num_classes = 10)
y_test <- to_categorical(cifar10$test$y, num_classes = 10)


N <- 48
idx_start <- (N - 32)/2 + 1
idx <- idx_start:(idx_start + 31)

# Select target image and format for prediction
#### Pretrained classifier requires 224x224
#### New classifier will take image size as is (NxN)

x_train_N <- array(0, dim = c(50000, N, N, 3))
x_train_N[, idx, idx, ] <- x_train[1:50000,,,]
y_train_N <- y_train[1:50000,]

x_test_N <- array(0, dim = c(10000, N, N, 3))
x_test_N[, idx, idx, ] <- x_test[1:10000,,,]
y_test_N <- y_test[1:10000,]

img_N <- x_test_N[1,,,]

input <- layer_input(shape = c(N, N, 3))

# Load VGG16 base (without classifier)
VGG16_base <- application_vgg16(
  include_top = FALSE,
  input_tensor = input
)

output <- VGG16_base$output %>%
  layer_flatten(input_shape = layer_input(shape = c(N, N, 512))) %>%
  layer_dense(units = 10, activation = "softmax")

model <- keras_model(input, output)

freeze_weights(model, from = 1, to = 'block5_conv3')

model %>% compile(
  loss = "categorical_crossentropy",
  optimizer = optimizer_rmsprop(lr = 0.0001, decay = 1e-6),
  metrics = "accuracy"
)

model %>% fit(
  x_train_N, y_train_N,
  batch_size = 12,
  epochs = 1,
  validation_data = list(x_test_N, y_test_N),
  shuffle = TRUE
)

predictions <- model %>% predict(img_N)

img_output <- model$output[,which.max(predictions)]

last_conv_layer <- model %>% get_layer("block5_conv3")
grads <- k_gradients(img_output, last_conv_layer$output)[[1]]

# finish making the heatmap
pooled_grads <- k_mean(grads, axis = c(1, 2, 3))
iterate <- k_function(list(model$input),
                      list(pooled_grads, last_conv_layer$output[1,,,]))
c(pooled_grads_value, conv_layer_output_value) %<-% iterate(list(img_N))
for (i in 1:512) {
  conv_layer_output_value[,,i] <-
    conv_layer_output_value[,,i] * pooled_grads_value[[i]]
}
heatmap <- apply(conv_layer_output_value, c(1,2), mean)
heatmap <- pmax(heatmap, 0)
heatmap <- heatmap / max(heatmap)

plot_heatmap(heatmap, width = N, height = N)
```

**Comment by skeydan** [comment]
Okay, I take it this is solved then!

**Comment by MirandaTilton** [comment]
One quick correction to the code posted by @srvanderplas:

'''
img_N <- x_test_N[1,,,]
dim(img_N) <- c(1, N, N, 3)
'''

Thanks for your help, @skeydan!

**Comment by skeydan** [comment]
You're welcome!

**Comment by pparshakov** [comment]
The code from @srvanderplas with correction of @MirandaTilberg produce 3*3 heatmap: 
![Rplot](https://user-images.githubusercontent.com/14823532/209477862-952326fd-5c2f-4a6a-96c9-f2a9900ffe3a.png)

Could you please help, what am I doin wrong?
